rclcpp/src/rclcpp/executors/executor_entities_collector.cpp

```diff
@@ -0,0 +1,281 @@
+
+  for (const auto & group_ptr : manually_added_groups_) {
+    groups.push_back(group_ptr);
+  }
+  for (auto const & group_ptr: automatically_added_groups_) {
+    groups.push_back(group_ptr);
+  }
+  return groups;
+}
+
+std::vector<rclcpp::CallbackGroup::WeakPtr>
+ExecutorEntitiesCollector::get_manually_added_callback_groups()
+{
+  std::vector<rclcpp::CallbackGroup::WeakPtr> groups;
+  std::lock_guard<std::mutex> guard{mutex_};
+  for (const auto & group_ptr : manually_added_groups_) {
+    groups.push_back(group_ptr);
+  }
+  return groups;
+}
+
+std::vector<rclcpp::CallbackGroup::WeakPtr>
+ExecutorEntitiesCollector::get_automatically_added_callback_groups()
+{
+  std::vector<rclcpp::CallbackGroup::WeakPtr> groups;
+  std::lock_guard<std::mutex> guard{mutex_};
+  for (auto const & group_ptr: automatically_added_groups_) {
+    groups.push_back(group_ptr);
+  }
+  return groups;
+}
+
+void
+ExecutorEntitiesCollector::update_collections(){
+  this->add_automatically_associated_callback_groups();
+  this->prune_invalid_nodes_and_groups();
+}
+
+std::shared_ptr<ExecutorNotifyWaitable>
+ExecutorEntitiesCollector::get_executor_notify_waitable()
+{
+  return this->notify_waitable_;
+}
+
+void
+ExecutorEntitiesCollector::add_callback_group_to_collection(rclcpp::CallbackGroup::SharedPtr group_ptr,
+                                                     CallbackGroupCollection & collection)
+{
+  std::atomic_bool & has_executor = group_ptr->get_associated_with_executor_atomic();
+  if (has_executor.exchange(true)) {
+    throw std::runtime_error("Callback group has already been added to an executor.");
+  }
+  auto iter = collection.insert(group_ptr);
+  if (iter.second == false) {
+    throw std::runtime_error("Callback group has already been added to this executor.");
+  }
+  this->notify_waitable_->add_guard_condition(group_ptr->get_notify_guard_condition().get());
+}
+
+void
+ExecutorEntitiesCollector::remove_callback_group_from_collection(rclcpp::CallbackGroup::SharedPtr group_ptr,
+                                                         CallbackGroupCollection & collection)
+{
+  rclcpp::CallbackGroup::WeakPtr weak_group_ptr(group_ptr);
+  auto iter = collection.find(weak_group_ptr);
+  if (iter != collection.end()) {
+    // Check that the group hasn't been orphaned from it's respective node
+    if (!group_ptr->has_valid_node()) {
+      throw std::runtime_error("Node must not be deleted before its callback group(s).");
+    }
+
+    // Disassociate the callback group with the executor
+    std::atomic_bool & has_executor = group_ptr->get_associated_with_executor_atomic();
+    has_executor.store(false);
+    collection.erase(iter);
+    this->notify_waitable_->remove_guard_condition(group_ptr->get_notify_guard_condition().get());
+  } else {
+    throw std::runtime_error("Callback group needs to be associated with executor.");
+  }
+}
+
+void
+ExecutorEntitiesCollector::add_automatically_associated_callback_groups()
```

Comment:
this function could take a list of weak nodes as input.
when we call it after `add_node` or `add_callback_group` we don't need to loop through all nodes, but only on the new one.

Reply:
I think that we need to check for the condition where we have an existing node and an automatically-associated callback group is added to that node, correct?